server/src/main/java/org/elasticsearch/cluster/coordination/stateless/StoreHeartbeatService.java

```diff
@@ -134,13 +138,22 @@ protected void doRun() throws Exception {
                 return;
             }
 
-            final var registerTerm = currentTermSupplier.getAsLong();
-            if (registerTerm == heartbeatTerm) {
-                heartbeatStore.writeHeartbeat(new Heartbeat(heartbeatTerm, absoluteTimeInMillis()), rerunListener);
-            } else {
-                assert heartbeatTerm < registerTerm;
-                listener.onResponse(registerTerm);
-            }
+            currentTermSupplier.accept(rerunListener.delegateFailure((delegate, registerTermOpt) -> {
+                if (registerTermOpt.isEmpty()) {
+                    rerunListener.onResponse(retryAfterTermReadFailureDelay);
```

Comment:
I think it's fine to not cap the max number of retries here, if we fail to write the heartbeat a few times some other node should take over.